lib/Mojo/WebSocketProxy/Dispatcher/Parser.pm

```diff
@@ -0,0 +1,102 @@
+package Mojo::WebSocketProxy::Dispatcher::Parser;
+
+use strict;
+use warnings;
+
+sub parse_req {
+    my ($c, $req_storage) = @_;
+
+    my $result;
+    my $args = $req_storage->{args};
+    if (ref $args ne 'HASH') {
+        # for invalid call, eg: not json
+        $req_storage->{args} = {};
+        $result = $c->wsp_error('error', 'BadRequest', 'The application sent an invalid request.');
+    }
+
+    $result = _check_sanity($c, $req_storage) unless $result;
+
+    return $result;
+}
+
+sub _check_sanity {
+    my ($c, $req_storage) = @_;
+
+    my @failed;
+    my $args   = $req_storage->{args};
+    my $config = $c->wsp_config->{config};
+
+    OUTER:
+    foreach my $k (keys %$args) {
+        if (not ref $args->{$k}) {
+            last OUTER if (@failed = _failed_key_value($k, $args->{$k}, $config->{skip_check_sanity}));
+        } else {
+            if (ref $args->{$k} eq 'HASH') {
+                foreach my $l (keys %{$args->{$k}}) {
+                    last OUTER
+                        if (@failed = _failed_key_value($l, $args->{$k}->{$l}, $config->{skip_check_sanity}));
+                }
+            } elsif (ref $args->{$k} eq 'ARRAY') {
+                foreach my $l (@{$args->{$k}}) {
+                    last OUTER if (@failed = _failed_key_value($k, $l, $config->{skip_check_sanity}));
+                }
+            }
+        }
+    }
+
+    if (@failed) {
+        $c->app->log->warn("Sanity check failed: " . $failed[0] . " -> " . ($failed[1] // "undefined"));
+        my $result = $c->wsp_error('sanity_check', 'SanityCheckFailed', 'Parameters sanity check failed.');
+        if (    $result->{error}
+            and $result->{error}->{code} eq 'SanityCheckFailed')
+        {
+            $req_storage->{args} = {};
+        }
+        return $result;
+    }
+    return;
+}
+
+sub _failed_key_value {
+    my ($key, $value, $skip_check_sanity) = @_;
+
+    if ($skip_check_sanity && $key =~ /$skip_check_sanity/) {
+        return;
+    } elsif (
+        $key !~ /^[A-Za-z0-9_-]{1,50}$/
```

Comment:
@evostrov  @basiliscos  If you mean why we have sanity check at all here, it was already there when I started editing it. Somebody think is safer to only whitelist these char set.

Reason why L74 is added is, the check in L66 doesn't allow Japanese characters. So I added this exception.

If you think the while sanity check doesn't make sense, I am not against it. But check with @tfoertsch @raunakkathuria  @kavehmz